### PR TITLE
Migrate to the latest Gradle Build Cache API

### PR DESCRIPTION
**Pull request by bamboo**
- Replace `OriginTaskExecutionMetadata` by `OriginMetadata`

### CODE REVIEW

**Comment by wolfs** [comment]
@bamboo: Why do you think that using `RootBuildCacheControllerRef` is a problem? The best way would probably be to use the `Gradle` scoped `BuildCacheController`.

For nested builds, the `RootBuildCacheControllerRef` would already be available before the `Gradle` object, so you could even cache settings scripts from there, right? I don't know if it is worth the effort though.

**Comment by bamboo** [comment]
@wolfs thanks for taking a look.

I was thinking the use might be problematic because of the call to `forNonRootBuild`, it's always protected by a call to `isSet` though so should be fine?

`ScriptCache` is a build service so it can't depend on `BuildCacheController` directly.

**Comment by wolfs** [comment]
@bamboo Yeah, it is protected, though it throws an exception. Do we always expect that the `RootBuildCacheControllerRef` is set when you query it? The I would just call `forNonRootBuild`, else maybe we should check `isSet` first.

**Comment by bamboo** [comment]
Thinking a litte more about this, having the Kotlin DSL call `forNonRootBuild` seems to imply more knowledge than what the Kotlin DSL currently has (is it really a non root build? does it matter?) but the tests pass and looking at the code it seems harmless so I think the PR is good to go as is.

**Comment by wolfs** [comment]
Probably the method shouldn't be called `forNonRootBuild` but something like `getExisting()`.

File: subprojects/provider/src/test/kotlin/org/gradle/kotlin/dsl/caching/LoadDirectoryTest.kt [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+
+        // then:
+        assert(
+            subject.load(packProducedBy(currentBuildInvocation))!!.metadata.isProducedByCurrentBuild
+        )
+
+        // and:
+        assert(
+            !subject.load(packProducedBy(previousBuildInvocation))!!.metadata.isProducedByCurrentBuild
+        )
```
Comment (JLLeitschuh) [inline subprojects/provider/src/test/kotlin/org/gradle/kotlin/dsl/caching/LoadDirectoryTest.kt]: Why not use `assertFalse`?

**Comment by lptr** [comment]
I don't think we should do this. Instead, Kotlin DSL should migrate to the new execution model we are going to use for cacheable work. I wouldn't start using that API in an external project, though, as we don't yet want to guarantee its stability.

Instead we should get Kotlin DSL merged into `gradle/gradle`. Once that's done, the build cache team can take care of migration to the new API.

**Comment by bamboo** [comment]
@lptr, thanks for the input.

How about keeping the PR only about the migration from the deprecated `OriginTaskExecutionMetadata` to `OriginMetadata`?

File: subprojects/provider/src/test/kotlin/org/gradle/kotlin/dsl/caching/LoadDirectoryTest.kt (hunk 1 above)
Comment (bamboo) [reply subprojects/provider/src/test/kotlin/org/gradle/kotlin/dsl/caching/LoadDirectoryTest.kt]: For symmetry.

Comment (JLLeitschuh) [reply subprojects/provider/src/test/kotlin/org/gradle/kotlin/dsl/caching/LoadDirectoryTest.kt]: 😕 Symmetry? Aren't `assertFalse` and `assertTrue` pretty standard?

Comment (bamboo) [reply subprojects/provider/src/test/kotlin/org/gradle/kotlin/dsl/caching/LoadDirectoryTest.kt]: The code is not using `assertTrue`.

Comment (bamboo) [reply subprojects/provider/src/test/kotlin/org/gradle/kotlin/dsl/caching/LoadDirectoryTest.kt]: Moved to `assertFalse` / `assertTrue`.

**Pull request merged**